src/snowflake/cli/api/metrics.py

```diff
@@ -86,7 +181,66 @@ def increment_counter(self, name: str, value: int = 1) -> None:
         else:
             self._counters[name] += value
 
+    @property
+    def current_span(self) -> Optional[CLIMetricsSpan]:
+        return self._in_progress_spans[-1] if len(self._in_progress_spans) > 0 else None
+
+    @contextmanager
+    def start_span(self, name: str) -> Iterator[CLIMetricsSpan]:
+        """
+        Starts a new span that tracks various metrics throughout its execution
+
+        Assumes that parent steps contain the entirety of their child steps
+        Parent steps are automatically populated as the most recently executed step, like a stack
+
+        Only one span can be active at a time; once one span finishes its parent becomes the active span
+
+        Spans are not emitted in telemetry if depth/total limits are exceeded
+
+        :raises CliMetricsInvalidUsageError: if the step name is empty
+        """
+        new_step = CLIMetricsSpan(
+            name=name,
+            start_time=time.monotonic() - self._monotonic_start,
+            parent=self.current_span,
+        )
+
+        if len(self._in_progress_spans) >= self.IN_PROGRESS_SPANS_DEPTH_LIMIT:
+            self.num_spans_past_depth_limit += 1
+        else:
+            self._in_progress_spans.append(new_step)
+
+        try:
+            yield new_step
+        except BaseException as err:
```

Comment:
```suggestion
        except Exception as err:
```
we don't want to catch `KeyboardInterrupt`, `SystemExit`, etc since they're not errors

Reply:
I think it *might* be helpful for something like `KeyboardInterrupt` if our users are stuck on a super long step and they consistently exit out, wdyt?